### PR TITLE
[SCSB-145] require Python 3.10

### LINKED ISSUE #2000: MNT: Unpin y_py for Python 3.9 when we can
Revert:

* #1999 

Blocked by:

* https://github.com/y-crdt/ypy/issues/114
* https://github.com/jupyterlab/jupyterlab/issues/13920

### PR DESCRIPTION
**Pull request by zacharyburnett**
resolves [SCSB-145](https://jira.stsci.edu/browse/SCSB-145)

propagate Astropy's deprecation of Python 3.9 to downstream packages

> [!NOTE]
> this PR was generated automatically by ``batchpr`` :robot:

Close #2000 

[🐱](https://jira.stsci.edu/browse/JDAT-4339)

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 [project]
 name = "jdaviz"
 description = "Astronomical data analysis development leveraging the Jupyter platform"
-requires-python = ">=3.9"
+requires-python = ">=3.10"
```
Comment (pllim) [inline pyproject.toml:4]: Only this line seems relevant. Is the rest some overzealous ruff gone wild?

**Review by pllim — CHANGES_REQUESTED** [review]
You also will have to update the test matrix and comb doc for any mention of Python 3.9 and update that as well.

File: pyproject.toml (hunk 1 above)
Comment (zacharyburnett) [reply pyproject.toml:4]: yes that looks like the case; I'll prune the other changes

File: pyproject.toml [hunk 2]
```diff
@@ -68,16 +68,16 @@ imviz = "jdaviz.configs.imviz"
 [project.optional-dependencies]
 all = [
     "opencv-python",
-    "pysiaf"
+    "pysiaf",
```
Comment (zacharyburnett) [inline pyproject.toml:71]:
```suggestion
    "pysiaf"
```

File: pyproject.toml [hunk 3]
```diff
@@ -68,16 +68,16 @@ imviz = "jdaviz.configs.imviz"
...
 ]
 test = [
     "pytest",
     "pytest-astropy",
-    "pytest-tornasync"
+    "pytest-tornasync",
```
Comment (zacharyburnett) [inline pyproject.toml:76]:
```suggestion
    "pytest-tornasync"
```

File: pyproject.toml [hunk 4]
```diff
@@ -68,16 +68,16 @@ imviz = "jdaviz.configs.imviz"
...
 ]
 docs = [
     "sphinx-astropy[confv2]>=1.9.1",
-    "sphinx_design"
+    "sphinx_design",
```
Comment (zacharyburnett) [inline pyproject.toml:80]:
```suggestion
    "sphinx_design"
```

File: pyproject.toml [hunk 5]
```diff
@@ -86,7 +86,7 @@ roman = [
 [build-system]
 requires = [
     "setuptools>=61.2",
-    "setuptools_scm"
+    "setuptools_scm",
```
Comment (zacharyburnett) [inline pyproject.toml:89]:
```suggestion
    "setuptools_scm"
```

File: pyproject.toml [hunk 6]
```diff
@@ -133,12 +133,10 @@ filterwarnings = [
     "ignore:.*With traitlets 4\\.1, metadata should be set using the \\.tag\\(\\) method:DeprecationWarning",
     "ignore:The unit 'Angstrom' has been deprecated in the VOUnit standard\\. Suggested.* 0\\.1nm\\.",
     "ignore:((.|\n)*)Sentinel is not a public part of the traitlets API((.|\n)*)",
-    "ignore:datetime\\.datetime\\.utcfromtimestamp:DeprecationWarning",  # asdf + dateutil<=2.8.2 + Python 3.12
+    "ignore:datetime\\.datetime\\.utcfromtimestamp:DeprecationWarning",
```
Comment (zacharyburnett) [inline pyproject.toml:136]:
```suggestion
    "ignore:datetime\\.datetime\\.utcfromtimestamp:DeprecationWarning",  # asdf + dateutil<=2.8.2 + Python 3.12
```

File: pyproject.toml [hunk 7]
```diff
@@ -133,12 +133,10 @@ filterwarnings = [
...
     "ignore::DeprecationWarning:glue",
     "ignore::DeprecationWarning:asteval",
     "ignore:::specutils.spectra.spectrum1d",
-    # Ignore numpy 2.0 warning, see https://github.com/astropy/astropy/pull/15495
-    # and https://github.com/scipy/scipy/pull/19275
     "ignore:.*numpy\\.core.*:DeprecationWarning",
```
Comment (zacharyburnett) [inline pyproject.toml:140]:
```suggestion
    # Ignore numpy 2.0 warning, see https://github.com/astropy/astropy/pull/15495
    # and https://github.com/scipy/scipy/pull/19275
    "ignore:.*numpy\\.core.*:DeprecationWarning",
```

**Comment by zacharyburnett** [comment]
Sorry for the extra changes! I used this ticket to test making batch PRs

File: .github/workflows/ci_workflows.yml [hunk 8]
```diff
@@ -51,10 +51,10 @@ jobs:
             toxposargs: --remote-data
             allow_failure: false
 
-          - name: OS X - Python 3.9
+          - name: OS X - Python 3.10
             os: macos-latest
-            python: 3.9
-            toxenv: py39-test
+            python: 3.10
```
Comment (pllim) [inline .github/workflows/ci_workflows.yml:56]: Need quotes.
```suggestion
            python: '3.10'
```

File: .github/workflows/predeps_workflows.yml [hunk 9]
```diff
@@ -32,8 +32,8 @@ jobs:
 
           - name: RC testing on OSX
             os: macos-latest
-            python: 3.9
-            toxenv: py39-test-predeps
+            python: 3.10
```
Comment (pllim) [inline .github/workflows/predeps_workflows.yml:35]:
```suggestion
            python: '3.10'
```

File: .github/workflows/publish.yml [hunk 10]
```diff
@@ -23,7 +23,7 @@ jobs:
 
     - uses: actions/setup-python@v5
       with:
-        python-version: 3.9
+        python-version: 3.10
```
Comment (pllim) [inline .github/workflows/publish.yml:26]:
```suggestion
        python-version: '3.10'
```

File: CHANGES.rst [hunk 11]
```diff
@@ -128,6 +128,8 @@ Specviz2d
 Other Changes and Additions
 ---------------------------
 
+- bump required Python version to 3.10 [#2757]
```
Comment (pllim) [inline CHANGES.rst:131]:
```suggestion
- Bump required Python version to 3.10. [#2757]
```

**Comment by pllim** [comment]
I think `tox.ini` needs to be updated too.

Also just in case you can grep for the following:

`grep -ri python . | grep "3\.9"`

`grep -r py39 .`

**Review by pllim — APPROVED** [review]
Sorry this took a while. I rebased and cleaned it up. I think this is good to merge.

**Comment by pllim** [comment]
Merging would need admin override because branch protection rules have changed.

After merge, admin needs to update branch protection rule for the OSX job.

**Comment by pllim** [comment]
CI failure is unrelated (#2811).

**Comment by pllim** [comment]
cc @larrybradley and @camipacifici since they asked about dropping Python 3.9 earlier.

**Comment by pllim** [comment]
Thanks, @zacharyburnett !

**Pull request merged**